main/src/library/Iterator.flix

```diff
@@ -811,96 +708,67 @@ namespace Iterator {
     ///
     /// The original iterator `iter` should *not* be reused.
     ///
-    @Lazy
-    pub def filterMapL(f: a -> Option[b], iter: Iterator[a, r]): Iterator[b, r] \ Write(r) =
-        let Iterator(done, next) = iter;
-
-        // Allocate a reference to hold the next element.
-        let cursor = ref None @ Scoped.regionOf(iter);
-
-        let done1 = () -> match deref cursor {
-            case None => match filterMapForward(f, done, next) {
-                case None    => true
-                case Some(x) =>
-                    // An element was found. Store it in the reference.
-                    cursor := Some(x);
-                    false
+    pub def filterMap(f: a -> Option[b] \ ef1, iter: Iterator[a, ef, r]): Iterator[b, ef and ef1, r] =
+        let Iterator(rc, iterF) = iter;
+        let step = x -> match x {
+            case Ans(a) => match (f(a)) {
+                case Some(a1) => Ans(a1)
+                case None     => Skip
             }
-            case Some(_) => false
+            case Skip   => Skip
+            case Done   => Done
         };
+        let iterF1 = () -> step(iterF());
+        Iterator(rc, iterF1)
 
-        let next1 = () -> match deref cursor {
-            case None    => bug!("Empty iterator")
-            case Some(x) =>
-                // An element is in the reference. Use it.
-                cursor := None;
-                x
-        };
-        Iterator(done1, next1)
-
     ///
-    /// Fast forwards the given iterator until the next element that produces
-    /// a `Some` given function `f`. Returns it if it exists, or `None`.
+    /// Returns the contents of `iter` as a list.
+    ///
+    /// Consumes the entire iterator.
     ///
-    /// This implementation returns `f(x)` as opposed to `forward` which returns `x`.
+    /// If any stage of the iterator is `Err(e)` then `Err(e)` is returned (first fail).
+    /// If the iterator has no errors then the result is wrapped with `Ok`.
     ///
-    def filterMapForward(f: a -> Option[b] \ ef, done: Done[r], next: Next[a, r]): Option[b] \ { ef, Read(r) } =
-        if (done())
-            None
-        else
-            match f(next()) {
-                case None    => filterMapForward(f, done, next)
-                case Some(b) => Some(b)
-            }
+    pub def toResultList(iter: Iterator[Result[e, a], ef, r]): Result[e, List[a]] \ { ef, Read(r) } =
```

Comment:
Do we have these functions elsewhere? I guess its useful.

Reply:
I think `toResultList` was something Jakob must have needed. I haven't seen it else where.

For error handling in general we might end up needing another Iterator module where the `Step` date type includes an `Error` constructor.

Reply:
I would delete it (if it is unused). It seems very irregular. 

If we want such functions we could add them back in a systematic way.